gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java

```diff
@@ -31,70 +37,307 @@
+      "This *does not* mean the time GPU was working solely on the buffer during this time.";
+  private static final String COMPOSITION_TOOLTIP =
+      "The time from when SurfaceFlinger latched on to the buffer and sent for composition " +
+      "to when it was sent to the display";
 
-  public boolean isEmpty() {
-    return buffers.isEmpty();
+  private List<Layer> layers;
+
+  private FrameInfo(List<Layer> layers) {
+    this.layers = layers;
   }
 
-  public int bufferCount() {
-    return buffers.size();
+  public int layerCount() {
+    return layers.size();
   }
 
-  public Iterable<Buffer> buffers() {
-    return Iterables.unmodifiableIterable(buffers);
+  public Iterable<Layer> layers() {
+    return Iterables.unmodifiableIterable(layers);
   }
 
   public static ListenableFuture<Perfetto.Data.Builder> listFrames(Perfetto.Data.Builder data) {
-    return transform(info(data.qe), frame -> data.setFrame(frame));
+    List<String> layerNames = Lists.newArrayList();
+    List<Layer> layers = Lists.newArrayList();
+    return transformAsync(data.qe.query(LAYER_NAME_QUERY), r -> {
+      r.forEachRow(($, row) -> {
+        layerNames.add(row.getString(0));
+      });
+      if (layerNames.isEmpty()) {
+        data.setFrame(FrameInfo.NONE);
+        return Futures.immediateFuture(data);
+      }
+      return transform(getLayers(data.qe, layers, layerNames), $ -> {
+        data.setFrame(new FrameInfo(layers));
+        return data;
+      });
+    });
   }
 
-  private static ListenableFuture<FrameInfo> info(QueryEngine qe) {
-    return
-        transform(qe.queries(FRAME_SLICE_QUERY), frame -> {
-            List<Buffer> buffers = Lists.newArrayList();
-            frame.forEachRow(($, r) -> {
-              buffers.add(new Buffer(r));
-            });
+  private static ListenableFuture<List<Layer>> getLayers(QueryEngine qe, List<Layer> layers,
+      List<String> layerNames) {
+    List<List<Event>> phases = Lists.newArrayList();
+    List<List<Event>> buffers = Lists.newArrayList();
+    // The group hierarchy looks like:
+    // > Layer 1
+    //   - DISPLAY
+    //   - APP
+    //   - Wait for GPU
+    //   - COMPOSITION
+    //   - > Buffers
+    //   -   - Buffer 1
+    //   -   - ...
+    //   -   - Buffer n
+    // ...
+    // > Layer m
+
+    // For every layer, add the four phases first, then add the buffers.
+    return transformAsync(createPhases(qe, phases, layerNames, 0), $1 -> {
+      return transform(createBuffers(qe, buffers, layerNames, 0), $2 -> {
+        for (int i = 0; i < layerNames.size(); i++) {
+          layers.add(new Layer(layerNames.get(i), buffers.get(i), phases.get(i)));
+        }
+        return layers;
+      });
+    });
+  }
+
+  private static ListenableFuture<List<List<Event>>> createPhases(QueryEngine qe,
+      List<List<Event>> phases, List<String> layerNames, int idx) {
+    String layerName = layerNames.get(idx);
+    String baseName = layerName.replaceAll("[^A-Za-z0-9]", "");
```

Comment:
What happens if you do not filter out non-alphanumeric characters?

Reply:
The layer names will usually have lots of dots(.) in them and sql will treat them as databases. For example, com.app.layer1 - com will be treated as a db and app as a table in it and so on.